src/session/session-utils.c

```diff
@@ -70,23 +63,74 @@ read_authorize_response (const char *what)
     }
 
+  /* make room for and add null terminator */
+  message = realloc (message, len + 1);
+  message[len] = '\0';
+  return (char *)message;
+}
+
+static const char *skip_quoted_value (const char *str)
+{
+  if (!*str || *str++ != '"')
+    errx (EX, "syntax error, expected quoted value");
+  while (*str && *str != '"')
+    str++;
+  if (*str != '"')
+    errx (EX, "syntax error, expected closing quote");
+  return str + 1;
+}
+
+char *get_authorize_key (const char *json, const char *key, bool required)
+{
   /*
-   * The authorize messages we receive always have an exact prefix and suffix:
+   * The authorize messages we receive always have this structure:
+   *
+   * \n{"command":"authorize","cookie":"NNN","response":"....","remote-peer":"..."}
    *
-   * \n{"command":"authorize","cookie":"NNN","response":"...."}
+   * We make many assumptions/restrictions here to keep the parser simple. In particular, no extra
+   * whitespace, everything is a string, no escaped ", and values don't contain commas.
    */
-  if (len <= auth_prefix_size + auth_response_size + auth_suffix_size ||
-      memcmp (message, auth_prefix, auth_prefix_size) != 0 ||
-      memcmp (message + auth_prefix_size, auth_response, auth_response_size) != 0 ||
-      memcmp (message + (len - auth_suffix_size), auth_suffix, auth_suffix_size) != 0)
+  if (!json || *json++ != '\n' || *json++ != '{')
+    errx (EX, "authorize message doesn't start with \\n{");
+
+  char *key_quoted;
+  int key_quoted_len = asprintf (&key_quoted, "\"%s\"", key);
+  if (key_quoted_len < 0)
+    errx (EX, "out of memory");
+
+  /* parse each "key":"value" pair until we find a matching key */
+  while (*json != '\0' && *json != '}')
     {
-      errx (EX, "didn't receive expected \"authorize\" message");
+      /* is this the key we are looking for? */
+      if (strncmp (json, key_quoted, key_quoted_len) == 0)
+        {
+          json += key_quoted_len;
+          if (*json++ != ':')
+              errx (EX, "authorize message syntax error, expected ':' after key");
+          /* check for value */
+          if (*json++ != '"')
+              errx (EX, "authorize message value isn't a string");
+
+          /* find end of value */
+          const char *value = json;
+          while (*json != '"')
+              json++;
+          char *res = strndup (value, json - value);
+          return res;
+        }
+
+      /* skip to next key */
+      json = skip_quoted_value (json);
+      if (*json++ != ':')
+          errx (EX, "authorize message syntax error, expected ':' after key");
+      json = skip_quoted_value (json);
+      if (*json++ != ',')
+        errx (EX, "authorize message missing comma");
     }
 
-  len -= auth_prefix_size + auth_response_size + auth_suffix_size;
-  memmove (message, message + auth_prefix_size + auth_response_size, len);
-  message[len] = '\0';
-  return (char *)message;
+  if (required)
+    errx (EX, "authorize message missing required key %s", key);
```

Comment:
I guess we don't have any backwards compatibility concerns here.  `cockpit-session` is only ever run from `cockpit-ws` on the same host, shipped in the same package.

Reply:
This isn't a behaviour change. It failed before as well, just with a different message:

```C
  if (len <= auth_prefix_size + auth_response_size + auth_suffix_size ||
      memcmp (message, auth_prefix, auth_prefix_size) != 0 ||
      memcmp (message + auth_prefix_size, auth_response, auth_response_size) != 0 ||
      memcmp (message + (len - auth_suffix_size), auth_suffix, auth_suffix_size) != 0)
    {
      errx (EX, "didn't receive expected \"authorize\" message");
    }
```
That's why I introduced the `required` flag. The error message got more specific, but missing `response` key was fatal before.